### ISSUE TITLE
Signing uploaded packages

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/conda-forge.github.io#129]
This should hopefully be pretty straightforward and would be a great asset to the users. Basically we should sign packages that we upload. Thus one can verify that the packages have not been tampered with in anyway before installing them.

**Comment by jakirkham** [comment]
cc @Cadair

**Comment by Cadair** [comment]
I have been thinking about this a fair bit recently, and looking into the ability to do package signing with PyPI and then thinking about what it means for conda. I put a lot of my thoughts down here: http://stuartmumford.uk/blog/verifying-pypi-and-conda-packages.html

I think the challenge with conda-forge is what exactly are you verifying? If there was a conda-forge PGP key which signed the binaries, you would be verifying that the package was the one actually built from the repo with the pass key in it.

Finally, I think what is a good idea is that we implement signature verification of the PyPI downloads. We can put a list of allowed pub keys in the recipie and do the check before we build the packages.
This would then chain with any signing of the conda packages, which would then be an implicit validation of the source we built from.

**Comment by kain88-de** [comment]
> Finally, I think what is a good idea is that we implement signature verification of the PyPI downloads. 

Isn't this already done with the hash check?

**Comment by Cadair** [comment]
@kain88-de The hash check is only designed to check that the file has not been corrupted in the download stage. Apart from the fact that the default md5 hash is really cryptographically weak, checking against a PGP signature verifies the uploader of the tarball to PyPI.

**Comment by kain88-de** [comment]
Thanks that answer helps. Are PyPi package required to be pgp signed?

**Comment by Cadair** [comment]
No, you can optionally upload a signature alongside the tarball.
